UART_driver/Core/Src/CircularBuffer.cpp

```diff
@@ -0,0 +1,50 @@
+#include "CircularBuffer.hpp"
+#include <cstring>
+
+CircularBuffer::CircularBuffer(uint8_t* buf, uint16_t size) {
+    this->buf = buf;
+    this->size = size;
+}
+
+uint8_t CircularBuffer::peek(uint8_t& res, uint16_t dist) {
+    if(dist + readPtr >= writePtr){
+        return 0;
+    }
+    res = buf[(readPtr + dist) % size];
+    return 1;
+}
+
+uint8_t CircularBuffer::read(uint8_t* res, uint16_t dist) {
+    if(dist + readPtr > writePtr) return 0;
+
+    if( (readPtr % size) + dist >= size ) {
+        // two memcpys needed
+        int dist_to_end = size - (readPtr % size);
```

Comment:
should this be int here? will the distance ever be neg? can we also try to use explicit size types as well? Seems like `size` is only a uint16_t so do we really ever need anything greater than that type?

In embedded we really try to use only the memory we need... also be specifying uint or int allows the reader to know what range of values to expect for that variable.